File: ActionUI/Common/ActionUILogger.swift
// Common/ActionUILogger.swift
import Foundation

/// Protocol for logging messages in the ActionUI library with specified severity levels.
/// Used for debugging, error reporting, and informational logging during view rendering and validation.
protocol ActionUILogger {
    /// Logs a message with the specified severity level.
    /// - Parameters:
    ///   - message: The message to log.
    ///   - level: The severity level of the message (error, warning, info, debug, or verbose).
    func log(_ message: String, _ level: Level)
}

/// Enum defining the severity levels for logging in the ActionUI library.
/// Uses Int raw values to enable filtering logs below a certain level (e.g., log only if level.rawValue <= maxLevel).
/// Lower values indicate higher severity.
enum Level: Int {
    /// Indicates a critical issue that may prevent normal operation (e.g., invalid JSON causing view rendering failure).
    case error = 1
    /// Indicates a non-critical issue that may affect functionality (e.g., missing optional property with a fallback).
    case warning = 2
    /// Indicates general information for debugging or tracking (e.g., view registration or state update).
    case info = 3
    /// Indicates detailed debugging information for developers (e.g., intermediate state changes or binding updates).
    case debug = 4
    /// Indicates exhaustive diagnostic information (e.g., every property validation or view construction step).
    case verbose = 5
}

File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Common/ActionUILogger.kt
package com.abracode.actionui.Common

import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.staticCompositionLocalOf

/**
 * CompositionLocal carrying the active [ActionUILogger] through the rendered
 * tree. `ActionUI.Render` provides it; element builders read it via
 * `LocalActionUILogger.current` to forward warnings (unknown colors, missing
 * SwiftUI-only alignment equivalents, etc.) without threading an explicit
 * logger parameter through every Composable signature.
 *
 * `staticCompositionLocalOf` because the logger rarely changes within a
 * rendered tree — subtree overrides aren't expected and the value is read
 * frequently.
 */
val LocalActionUILogger: ProvidableCompositionLocal<ActionUILogger> =
    staticCompositionLocalOf { ConsoleLogger() }

/**
 * Logger interface for the ActionUI library with leveled messages.
 *
 * Mirror of Swift `ActionUILogger` protocol in
 * `ActionUI/Common/ActionUILogger.swift`. Implementations should be safe to
 * invoke from any thread the renderer happens to run on.
 *
 * Used for debugging, error reporting, and informational logging during view
 * rendering and validation.
 */
interface ActionUILogger {
    /**
     * Logs a message with the specified severity level.
     *
     * @param message The message to log.
     * @param level The severity level of the message.
     */
    fun log(message: String, level: LoggerLevel)
}

/**
 * Severity levels for ActionUI logging. Lower [rawValue] means higher severity.
 *
 * Mirror of Swift `LoggerLevel` enum. The raw values must agree across both
 * platforms so JSON-driven configuration ("log up to level 3") behaves the
 * same way on Android and Apple targets.
 */
enum class LoggerLevel(val rawValue: Int) {
    /** Critical issue that may prevent normal operation (e.g., invalid JSON). */
    error(1),
    /** Non-critical issue that may affect functionality (e.g., dropped key). */
    warning(2),
    /** General information for debugging or tracking. */
    info(3),
    /** Detailed debugging information. */
    debug(4),
    /** Exhaustive diagnostic information. */
    verbose(5);
}
